gstd/src/exec/async.rs

```diff
@@ -0,0 +1,69 @@
+// This file is part of Gear.
+
+// Copyright (C) Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{async_runtime, msg, MessageId};
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use gcore::exec;
+
+/// Delays message execution in asynchronous way for the specified number of
+/// blocks. It works pretty much like the `exec::wait_for` function, but allows
+/// to continue execution after the delay in the same handler. It is worth
+/// mentioning that the program state gets persisted inside the call, and the
+/// execution resumes with potentially different state.
+pub fn delay_for(block_count: u32) -> impl Future<Output = ()> {
+    MessageDelayFuture::new(msg::id(), exec::block_height().saturating_add(block_count))
+}
+
+struct MessageDelayFuture {
+    msg_id: MessageId,
+    till_block_number: u32,
+}
+
+impl MessageDelayFuture {
+    fn new(msg_id: MessageId, till_block_number: u32) -> Self {
+        Self {
+            msg_id,
+            till_block_number,
+        }
+    }
+}
+
+impl Future for MessageDelayFuture {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let current_block_number = exec::block_height();
+
+        async_runtime::locks().remove(self.msg_id, self.msg_id);
+        if current_block_number < self.till_block_number {
+            async_runtime::locks().lock(
+                self.msg_id,
+                self.msg_id,
+                async_runtime::Lock::exactly(self.till_block_number - current_block_number)
+                    .expect("Unreachable: never fails with block count > 0"),
+            );
+            Poll::Pending
+        } else {
+            Poll::Ready(())
+        }
```

Comment:
Comparing `current_block_number < self.till_block_number` looks more robust to me, as it works when some blocks are skipped.

Reply:
Ah, sorry, misunderstood, your code is similar. But I don't see the difference except that `block_count` may be equal to zero in your case.

Reply:
We need the `self.till_block_number - current_block_number` to be exactly > 0 otherwise the `Lock::exactly` returns an error. That's why it is safe to write `.expect("Unreachable: never fails with block count > 0"),`